### PR TITLE
StatefulSet: Cleanup the complex defer function updating the status

### PR DESCRIPTION
**Pull request by gjkim42**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:

ref: https://github.com/kubernetes/kubernetes/pull/109694#discussion_r979974744

In the long term, the complex defer function makes the code harder to maintain as code after it should take that into account. This removes the complex defer function updating the status of a statefulset.

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

I'll make cherry-pick PRs containing both https://github.com/kubernetes/kubernetes/pull/109694 and this if you are ok with that to address https://github.com/kubernetes/kubernetes/issues/112010.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

/cc @soltysh @atiratree

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set_control.go [hunk 1]
```diff
@@ -325,23 +342,6 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		// If the ordinal could not be parsed (ord < 0), ignore the Pod.
 	}
 
-	// make sure to update the latest status even if there is an error later
-	defer func() {
```
Comment (smarterclayton) [inline pkg/controller/statefulset/stateful_set_control.go:329]: Do we have sufficient unit tests to cover this at a branch level?

**Comment by smarterclayton** [comment]
Looks correct to me.

As I was reviewing this, I did frown at bit at an existing line (not yours):

https://github.com/kubernetes/kubernetes/pull/112737/files#diff-a1e25eb50e4b2aed8947c2caa5d6d1ddfef1a5f2f95fdf71f2388c657823ee13R572

Decrementing current replicas if the delete fails seems incorrect.  If I read correctly we would actually have N+1 replicas when delete fails, and so we would be reporting an incorrect number. If this is correct, at minimum we should have a comment why this is safe.

Other than the question about unit test coverage, I looked through the branches and the change appears to be correct.  

Is this the only caller of that method?

File: pkg/controller/statefulset/stateful_set_control.go (hunk 1 above)
Comment (smarterclayton) [reply pkg/controller/statefulset/stateful_set_control.go:329]: I assume all the unit tests are on UpdateStatefulSet, not the nested method?

Comment (gjkim42) [reply pkg/controller/statefulset/stateful_set_control.go:329]: > Is this the only caller of that method?

Yes.

> I assume all the unit tests are on UpdateStatefulSet, not the nested method?

Yes, they are.

---

https://github.com/kubernetes/kubernetes/pull/109694 (the original PR) merged with some integration tests without unit tests. I'll add unit some tests for this.

File: pkg/controller/statefulset/stateful_set_control_test.go [hunk 2]
```diff
@@ -2101,6 +2101,72 @@ func TestStatefulSetAvailability(t *testing.T) {
...
+			hasSyncErr:   true,
+			hasStatusErr: false,
+			expectedErr:  syncErr,
+		},
+		{
+			desc:         "status error",
+			hasSyncErr:   false,
+			hasStatusErr: true,
+			expectedErr:  statusErr,
+		},
```
Comment (atiratree) [inline pkg/controller/statefulset/stateful_set_control_test.go:2135]: and also the sync + status error case?

**Comment by atiratree** [comment]
> Decrementing current replicas if the delete fails seems incorrect. If I read correctly we would actually have N+1 replicas when delete fails, and so we would be reporting an incorrect number. If this is correct, at minimum we should have a comment why this is safe.

It seems to me that this should not decrement upon error as well. @smarterclayton @gjkim42 Do we want to address that in this PR or a new one?

**Comment by gjkim42** [comment]
> It seems to me that this should not decrement upon error as well. @smarterclayton @gjkim42 Do we want to address that in this PR or a new one?

you meant here right?

https://github.com/kubernetes/kubernetes/blob/5301d921505971504867f88db9597c6d8e3e1957/pkg/controller/statefulset/stateful_set_control.go#L564-L574

I think that should be addressed by the new PR as that is an integrity issue.

File: pkg/controller/statefulset/stateful_set_control.go [hunk 3]
```diff
@@ -105,8 +105,28 @@ func (ssc *defaultStatefulSetControl) performUpdate(
...
+		klog.V(4).InfoS("Updated status", "statefulSet", klog.KObj(set),
+			"replicas", currentStatus.Replicas,
+			"readyReplicas", currentStatus.ReadyReplicas,
+			"currentReplicas", currentStatus.CurrentReplicas,
+			"updatedReplicas", currentStatus.UpdatedReplicas)
+	}
+
+	switch {
+	case err != nil && statusErr != nil:
+		klog.V(4).ErrorS(statusErr, "Could not update status", "statefulSet", klog.KObj(set))
```
[the remark this answers is not in the document]
Comment (gjkim42) [reply pkg/controller/statefulset/stateful_set_control.go]: I think most ErrorS should be V(0), fixed it.

**Comment by gjkim42** [comment]
> Do you want to give it a go, or should I?

I don't, it is ok if you do.

**Comment by atiratree** [comment]
@gjkim42 ow, we missed this:


https://github.com/kubernetes/kubernetes/blob/b4156ea47b3caf0088c6316ecd89afe715830807/pkg/controller/statefulset/stateful_set_control.go#L650

should return status as well

**Review by soltysh — APPROVED** [review]
/lgtm
/approve

**Comment by atiratree** [comment]
/hold

@soltysh I think you might have missed my last minute comment about returning status https://github.com/kubernetes/kubernetes/pull/112737#issuecomment-1305928531

It is just one line change. @gjkim42

**Comment by gjkim42** [comment]
> @gjkim42 ow, we missed this:
> 
> https://github.com/kubernetes/kubernetes/blob/b4156ea47b3caf0088c6316ecd89afe715830807/pkg/controller/statefulset/stateful_set_control.go#L650
> 
> should return status as well

@atiratree 
Thanks, done.

**Comment by atiratree** [comment]
/hold cancel
@gjkim42 thank you

**Comment by atiratree** [comment]
> Decrementing current replicas if the delete fails seems incorrect. If I read correctly we would actually have N+1 replicas when delete fails, and so we would be reporting an incorrect number. If this is correct, at minimum we should have a comment why this is safe.

@smarterclayton @gjkim42  FYI, I have created an issue and a PR that fixes that 

- https://github.com/kubernetes/kubernetes/issues/113833 
- https://github.com/kubernetes/kubernetes/pull/113834

**Pull request merged**